base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -838,6 +847,52 @@ where B: BlockchainBackend
         db.fetch_mmr_size(tree)
     }
 
+    pub fn get_validator_nodes_mr(&self) -> Result<Vec<u8>, ChainStorageError> {
+        println!("[+] get_validator_nodes_mr");
+        let validator_nodes = self.fetch_active_validator_nodes()?;
+        println!("validator_nodes {:?}", validator_nodes);
+        let mut hashes = validator_nodes
+            .iter()
+            .map(|validator_node| validator_node.shard_key.to_vec())
+            .collect::<Vec<Vec<u8>>>();
+        println!("hashes {:?}", hashes);
+        let x = hashes.len();
+        if x == 0 {
+            // Empty tree return 0
+            return Ok(vec![0; 32]);
+        }
+        let prev_power_of_two = (x + 1).next_power_of_two() / 2;
+        let mut over = x - prev_power_of_two;
+        let mut i = prev_power_of_two - over;
+        let mut j = i;
+        while over > 0 {
```

Comment:
It's better to use the MerkleMountainRange and just add all  of them into it as leaves. See how TransactionInputs are created